arangod/RocksDBEngine/RocksDBVectorIndex.cpp

```diff
@@ -0,0 +1,463 @@
+      TRI_ASSERT(opts.prefix_same_as_start);
+    });
+
+    _rocksdbKey.constructVectorIndexValue(_index->objectId(), _listNumber);
+    _it->Seek(_rocksdbKey.string());
+  }
+
+  [[nodiscard]] bool is_available() const override {
+    return _it->Valid() && _it->key().starts_with(_rocksdbKey.string());
+  }
+
+  void next() override { _it->Next(); }
+
+  std::pair<faiss::idx_t, const uint8_t*> get_id_and_codes() override {
+    auto const docId = RocksDBKey::indexDocumentId(_it->key());
+    TRI_ASSERT(_codeSize == _it->value().size());
+    const auto* value = reinterpret_cast<const uint8_t*>(_it->value().data());
+    return {static_cast<faiss::idx_t>(docId.id()), value};
+  }
+
+ private:
+  RocksDBKey _rocksdbKey;
+  arangodb::RocksDBVectorIndex* _index = nullptr;
+
+  std::unique_ptr<rocksdb::Iterator> _it;
+  std::size_t _listNumber;
+  std::size_t _codeSize;
+};
+
+struct RocksDBInvertedLists : faiss::InvertedLists {
+  RocksDBInvertedLists(RocksDBVectorIndex* index, LogicalCollection* collection,
+                       transaction::Methods* trx,
+                       RocksDBMethods* rocksDBMethods,
+                       rocksdb::ColumnFamilyHandle* cf, std::size_t nlist,
+                       size_t codeSize)
+      : InvertedLists(nlist, codeSize),
+        _index(index),
+        _collection(collection),
+        _trx(trx),
+        _rocksDBMethods(rocksDBMethods),
+        _cf(cf) {
+    use_iterator = true;
+    assert(status.ok());
+  }
+
+  std::size_t list_size(std::size_t /*listNumber*/) const override {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_SHUTTING_DOWN,
+                                   "faiss list_size not supported");
+  }
+
+  const std::uint8_t* get_codes(std::size_t /*listNumber*/) const override {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_SHUTTING_DOWN,
+                                   "faiss get_codes not supported");
+  }
+
+  const faiss::idx_t* get_ids(std::size_t /*listNumber*/) const override {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_SHUTTING_DOWN,
+                                   "faiss get_ids not supported");
+  }
+
+  size_t add_entries(std::size_t listNumber, std::size_t nEntry,
+                     const faiss::idx_t* ids,
+                     const std::uint8_t* code) override {
+    for (std::size_t i = 0; i < nEntry; i++) {
+      RocksDBKey rocksdbKey;
+      // Can we get away with saving LocalDocumentID in key, maybe we need to
+      // the use IdSelector in options during search
+      auto const docId = LocalDocumentId(*(ids + i));
+      rocksdbKey.constructVectorIndexValue(_index->objectId(), listNumber,
+                                           docId);
+
+      auto const value = RocksDBValue::VectorIndexValue(
+          reinterpret_cast<const char*>(code + i * code_size), code_size);
+      auto status =
+          _rocksDBMethods->PutUntracked(_cf, rocksdbKey, value.string());
+
+      TRI_ASSERT(status.ok());
```

Comment:
Ah yes, that is true, we will have to throw here because there is no safe way to return to normal execution since faiss is in control now. And catch in insert and handle it normally.